internal/bazel_integration_test/bazel_integration_test.bzl

```diff
@@ -0,0 +1,122 @@
+            TMPL_bazel_binary_workspace = ctx.attr.bazel_binary.label.workspace_name,
+            TMPL_bazel_args = ", ".join(["'%s'" % s for s in ctx.attr.bazel_args]),
+            TMPL_repositories = ", ".join(["'%s': '%s'" % (ctx.attr.repositories[f], _file_to_manifest_path(ctx, f.files.to_list()[0])) for f in ctx.attr.repositories]),
+            TMPL_bazelrc_append = ctx.attr.bazelrc_append,
+            TMPL_bazelrc_imports = ", ".join(["'%s': '%s'" % (ctx.attr.bazelrc_imports[f], _file_to_manifest_path(ctx, f.files.to_list()[0])) for f in ctx.attr.bazelrc_imports]),
+            TMPL_npm_packages = ", ".join(["'%s': '%s'" % (ctx.attr.npm_packages[f], _file_to_manifest_path(ctx, f.files.to_list()[0])) for f in ctx.attr.npm_packages]),
+            TMPL_package_json_replacements = ", ".join(["'%s': '%s'" % (f, ctx.attr.package_json_replacements[f]) for f in ctx.attr.package_json_replacements]),
+        ),
+    )
+
+    # Test executable is a shell script that runs ctx.executable._test_runner
+    # and passes it the generated configuration file
+    ctx.actions.write(
+        output = ctx.outputs.executable,
+        is_executable = True,
+        content = """#!/usr/bin/env bash
+# Immediately exit if any command fails.
+set -e
+
+if [ -e "$RUNFILES_MANIFEST_FILE" ]; then
+  while read line; do
+    declare -a PARTS=($line)
+    if [ "${{PARTS[0]}}" == "{TMPL_test_runner}" ]; then
+      readonly TEST_RUNNER=${{PARTS[1]}}
+    elif [ "${{PARTS[0]}}" == "{TMPL_args}" ]; then
+      readonly ARGS=${{PARTS[1]}}
+    fi
+  done < $RUNFILES_MANIFEST_FILE
+else
+  readonly TEST_RUNNER=../{TMPL_test_runner}
+  readonly ARGS={TMPL_args}
+fi
+
+$TEST_RUNNER $ARGS "$@"
+""".format(
+            TMPL_test_runner = _file_to_manifest_path(ctx, ctx.executable._test_runner),
+            TMPL_args = _file_to_manifest_path(ctx, config),
+        ),
+    )
+
+    runfiles = [config] + ctx.files.bazel_binary + ctx.files.workspace_files + ctx.files.repositories + ctx.files.bazelrc_imports + ctx.files.npm_packages
+
+    return [DefaultInfo(
+        runfiles = ctx.runfiles(files = runfiles).merge(ctx.attr._test_runner[DefaultInfo].data_runfiles),
+        executable = ctx.outputs.executable,
+    )]
+
+BAZEL_INTEGRATION_TEST_ATTRS = {
+    "bazel_args": attr.string_list(
+        default = ["test", "..."],
+    ),
+    "bazel_binary": attr.label(),
+    "bazelrc_append": attr.string(),
+    "bazelrc_imports": attr.label_keyed_string_dict(
+        allow_files = True,
+    ),
+    "npm_packages": attr.label_keyed_string_dict(),
+    "package_json_replacements": attr.string_dict(),
+    "repositories": attr.label_keyed_string_dict(),
+    "workspace_files": attr.label(),
+    "workspace_under_test": attr.string(),
+    "_test_runner": attr.label(
+        executable = True,
+        cfg = "host",
+        default = Label("@build_bazel_rules_nodejs//internal/bazel_integration_test:test_runner"),
+    ),
+}
+
+bazel_integration_test = rule(
```

Comment:
You might consider wrapping this in [a macro](https://github.com/bazelbuild/bazel-integration-testing/blob/master/go/bazel_integration_test.bzl#L44) and creating 1 test for each of the various versions of Bazel that are officially supported by rules_nodejs. It's good to get coverage for older Bazels, but it does come at the cost of a slower build. 

If you're not interested in testing anything other than the latest 3 versions

```
load("@com_github_bazelbuild_bazel_integration_testing//tools:common.bzl", "GET_LATEST_BAZEL_VERSIONS")
```

`GET_LATEST_BAZEL_VERSIONS` also takes an optional param, that defaults to 3, that is the number of minor versions to test and reduces the list to the latest patch version of each. So, for example, if 0.30.3, 0.30.2, 0.30.1, 0.30.0, 0.29.0, 0.28.1, and 0.28.0 are around, it'll give you 0.30.3, 0.29.0, and 0.28.1.

Reply:
we thought about that, there are currently no old bazels we support but when the first LTS one comes out we should test that we support it following the schedule Bazel has for support themselves